### PR TITLE
Fix ControllerRevision e2e test flake

### PR DESCRIPTION
**Pull request by riaankleinhans**
**What type of PR is this?**
/bug

**What this PR does / why we need it:**
The #110122  test is currently flaking badly and this PR will fix the issue.

**Testgrid Link:** [Link](https://testgrid.k8s.io/sig-apps#gce-serial&width=20&include-filter-by-regex=should.manage.the.lifecycle.of.a.ControllerRevision)

**Special notes for your reviewer:**
This is needed ASAP to address the flake

**Does this PR introduce a user-facing change?:**
```
NONE

```

**Release note:**
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:**
```
NONE

```

/sig testing
/sig architecture
/area conformance

### CODE REVIEW

**Comment by Riaankl** [comment]
/assign @kow3ns @soltysh @aojea 

Appoligies for assigning the whole world to this PR. Would like to get the flake addressed ASAP.
TIA

File: test/e2e/apps/controller_revision.go
Comment (aojea) [inline test/e2e/apps/controller_revision.go:157]: asking without  context, why 1 and why this fix the flake?

Comment (heyste) [reply test/e2e/apps/controller_revision.go:157]: AIUI the test currently fails because the hash for the daemonset `initialRevision.Data` is the same as the current running daemonset. Changing the spec for `TerminationGracePeriodSeconds` will mean that there is a different hash when the ControllerRevision is created. The number just needs to be different from the default value of 30. Most tests if I'm changing this value I prefer using 1.

**Comment by aojea** [comment]
/kind flake

@smarterclayton @soltysh can you PTAL?
https://github.com/kubernetes/kubernetes/pull/111075#discussion_r918330125

**Comment by aojea** [comment]
> **What this PR does / why we need it:** The #110122 test is currently flaking badly and this PR will fix the issue.
> 
> **Testgrid Link:** [Link](https://testgrid.k8s.io/sig-apps#gce-serial&width=20&include-filter-by-regex=should.manage.the.lifecycle.of.a.ControllerRevision)

but this is not flake, this is permanently failing

/kind failing-test
/remove-kind flake

**Comment by aojea** [comment]
we should merge the presubmit optional job and test before merge to avoid more problems

https://github.com/kubernetes/test-infra/pull/26556

File: test/e2e/apps/controller_revision.go
Comment (krmayankk) [reply test/e2e/apps/controller_revision.go:157]: how was this working earlier ?

Comment (heyste) [reply test/e2e/apps/controller_revision.go:157]: The current test passes without any issues when using a kind cluster but when It's checked with local-up-cluster it has a hard failure. I don't know the internals enough to know what's the difference between those two cluster setups.

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

**Comment by aojea** [comment]
> /priority critical-urgent

we can test before merge

/test pull-kubernetes-e2e-gce-ubuntu-containerd-serial

/hold

I'm just holding to watch the result, anybody here feel free to unhold once we check the test is not failing

File: test/e2e/apps/controller_revision.go
Comment (aojea) [reply test/e2e/apps/controller_revision.go:157]: the main difference on kind and gce clusters use to be the latency, kind use to be much faster ...

**Comment by Riaankl** [comment]
> > /priority critical-urgent
> 
> we can test before merge
> 
> /test pull-kubernetes-e2e-gce-ubuntu-containerd-serial
> 
> /hold
> 
> I'm just holding to watch the result, anybody here feel free to unhold once we check the test is not failing

@aojea I will check and retest as needed.

**Comment by Riaankl** [comment]
@aojea 
[pull-kubernetes-e2e-gce-ubuntu-containerd-serial](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/111075/pull-kubernetes-e2e-gce-ubuntu-containerd-serial/1546808666152243200/) log show:
````
I0712 11:23:34.739] Jul 12 11:23:34.739: INFO: DaemonSet pods can't tolerate node e2e-9fed553a7a-ac48a-master with taints [{Key:node-role.kubernetes.io/master Value: Effect:NoSchedule TimeAdded:<nil>} {Key:node.kubernetes.io/unschedulable Value: Effect:NoSchedule TimeAdded:<nil>}], skip checking this node
I0712 11:23:34.782] Jul 12 11:23:34.782: INFO: Number of nodes with available pods controlled by daemonset daemon-set: 3
I0712 11:23:34.782] Jul 12 11:23:34.782: INFO: Number of running nodes: 3, number of available pods: 3 in daemonset daemon-set
I0712 11:23:34.949] STEP: Update daemon pods image. 07/12/22 11:23:34.949
I0712 11:23:35.041] STEP: Check that daemon pods images are updated. 07/12/22 11:23:35.041
I0712 11:23:35.083] Jul 12 11:23:35.083: INFO: Wrong image for pod: daemon-set-4v47k. Expected: registry.k8s.io/e2e-test-images/agnhost:2.39, got: registry.k8s.io/e2e-test-images/httpd:2.4.38-2.
I0712 11:23:35.084] Jul 12 11:23:35.083: INFO: Wrong image for pod: daemon-set-n445h. Expected: registry.k8s.io/e2e-test-images/agnhost:2.39, got: registry.k8s.io/e2e-test-images/httpd:2.4.38-2.
```
Have not failed yet.

**Comment by Riaankl** [comment]
/unhold

All tests pass OK
Thanks for all the reviews and related work!!

**Comment by Riaankl** [comment]
/test pull-kubernetes-verify-govet-levee

Unrleated 
```
Test started today at 5:20 AM failed after 15m3s. (more info)

There are no nodes that your pod can schedule to - check your requests, tolerations, and node selectors (0/104 nodes are available: 101 Insufficient cpu, 13 Insufficient memory, 3 node(s) had taint {dedicated: greenhouse}, that the pod didn't tolerate.)
```

**Comment by Riaankl** [comment]
Test grid is running green since the fix was merged. Thanks everyone!
image

File: test/e2e/apps/controller_revision.go
Comment (atiratree) [reply test/e2e/apps/controller_revision.go:157]: Results of the investigation:

The hash is not just computed from the configured values (eg. what you would see in your struct initialization or in the generated YAML), but from all the fields in the DaemonSet struct ([hash.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/util/hash/hash.go)). So if any field changes you would get a different hash even though the final trimmed outputed YAML would stay the same.

This is basically what happened - it was cause by removal  https://github.com/kubernetes/kubernetes/pull/109602 of the `ZZZ_DeprecatedClusterName`  field from object metadata in master (1.25).

So if you connect to a bit older cluster running 1.24 (kind, etc) and you compute a hash from a DaemonSet ([controller_revision.go#L158](https://github.com/ii/kubernetes/blob/6c320b2991c5a9a94b1ac5a618dfec073b584d45/test/e2e/apps/controller_revision.go#L158)) it will differ from the cluster generated hash because the daemon_controller still has has the `ZZZ_DeprecatedClusterName` field, but the client (e2e test) does not have it anymore.

If you connect against the cluster of the same version (eg. `local-up-cluster.sh`) the hashes generated by both sides will be identical and indeed we need this fix that changes the DaemonSet struct a bit to get a different hash for a new revision.

**Pull request merged**
